[DLM] Introduce default rollover cluster setting & expose it via APIs

gmarouli opened the pull request:
In this PR we are introducing the following:
- A cluster setting that configures the default rollover used by DLM
- Introduces a new query param, namely `include_defaults` for `GET /data_streams`,  `GET /_index_template` and `GET /_component_template`, this will expose default configuration related to these endpoints.
- Exposes the default rollover configuration currently in place when the `include_defaults`
 parameter is true in the above endpoints
 - Uses the rollover configuration from the cluster setting in DLM.

Part of:  https://github.com/elastic/elasticsearch/issues/93596

**Docs preview:**

- [get data streams](https://elasticsearch_94240.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/indices-get-data-stream.html)
- [get index templates](https://elasticsearch_94240.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/indices-get-template.html)
- [get component templates](https://elasticsearch_94240.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/getting-component-templates.html)

gmarouli committed:
New RolloverConfiguration abstracts the RolloverAction configuration

gmarouli committed:
Use RolloverConfiguration in RolloverAction

gmarouli committed:
Use RolloverConfiguration in WaitForRolloverReadyStep

gmarouli committed:
Rename RolloverConfiguration to RolloverConditions

gmarouli committed:
Move RolloverConfiguration to a more appropriate package

gmarouli committed:
fix format

gmarouli committed:
fix format

gmarouli committed:
Use existing conditions

gmarouli committed:
Rework RolloverConditions to use a Map of Conditions

gmarouli committed:
Polish

gmarouli committed:
Merge branch 'main' into dlm-refactor-rollover

gmarouli committed:
Fix tests

gmarouli committed:
fix format

gmarouli committed:
Fix random RolloverConditions generation

gmarouli committed:
Fix random RolloverConditions generation

gmarouli committed:
Introduce rollover conditions setting parser

gmarouli committed:
Introduce default rollover condition setting

gmarouli committed:
Merge with main

gmarouli committed:
Extend toXContent of DataLifecycle to be able to add the RolloverConditions

gmarouli committed:
Extend toXContent of templates add the RolloverConditions

gmarouli committed:
Expose default RolloverConditions in GetComponentTemplateAction

gmarouli committed:
Expose default RolloverConditions in GetComposableIndexTemplateAction

gmarouli committed:
Expose default RolloverConditions in GetDataStreamAction

gmarouli committed:
Merge branch 'main' into dlm-rollover-cluster-setting

gmarouli committed:
Add yaml tests

gmarouli committed:
Polish

gmarouli committed:
Fix test

gmarouli committed:
Add validation

gmarouli committed:
Use the rollover cluster setting in the DataLifecycleService

gmarouli committed:
Extend docs to include new flag

gmarouli committed:
Fix typo, '_' to '-'

gmarouli committed:
Change position of preview macro

gmarouli committed:
Fix alignment

gmarouli committed:
Try to inline preview

gmarouli committed:
Merge with main

gmarouli committed:
Fix docs

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

gmarouli committed:
Change preview format.

andreidan commented on server/src/main/java/org/elasticsearch/action/admin/indices/template/get/TransportGetComposableIndexTemplateAction.java:
```diff
@@ -50,6 +54,7 @@ public TransportGetComposableIndexTemplateAction(
             GetComposableIndexTemplateAction.Response::new,
             ThreadPool.Names.SAME
         );
+        clusterSettings = DataLifecycle.isEnabled() ? clusterService.getClusterSettings() : null;
```
I think we could have the cluster settings non-null and add the DataLifecycle.isEnabled() check only when using them (to avoid potential NPEs)

andreidan commented on server/src/main/java/org/elasticsearch/action/admin/indices/template/get/GetComposableIndexTemplateAction.java:
```diff
@@ -98,14 +119,26 @@ public static class Response extends ActionResponse implements ToXContentObject
+            this.rolloverConditions = null;
+        }
+
+        public Response(Map<String, ComposableIndexTemplate> indexTemplates, RolloverConditions rolloverConditions) {
```
`@Nullable` for rolloverConditions

gmarouli replied:
Done

andreidan commented on server/src/main/java/org/elasticsearch/action/admin/indices/template/get/GetComposableIndexTemplateAction.java:
```diff
@@ -98,14 +119,26 @@ public static class Response extends ActionResponse implements ToXContentObject
         public static final ParseField INDEX_TEMPLATE = new ParseField("index_template");
 
         private final Map<String, ComposableIndexTemplate> indexTemplates;
+        private final RolloverConditions rolloverConditions;
```
`@Nullable`

gmarouli replied:
Done

andreidan commented on server/src/main/java/org/elasticsearch/action/datastreams/GetDataStreamAction.java:
```diff
@@ -289,30 +325,49 @@ public int hashCode() {
+            this(dataStreams, null);
+        }
+
+        public Response(List<DataStreamInfo> dataStreams, RolloverConditions rolloverConditions) {
```
@Nullable for rolloverConditions

gmarouli replied:
Ah, good one! Done

andreidan commented on server/src/main/java/org/elasticsearch/action/datastreams/GetDataStreamAction.java:
```diff
@@ -289,30 +325,49 @@ public int hashCode() {
             return dataStreams;
         }
 
+        public RolloverConditions getRolloverConditions() {
```
`@Nullable`

andreidan commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverConditions.java:
```diff
@@ -256,6 +257,60 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+        try {
+            return Long.parseLong(sValue);
+        } catch (NumberFormatException e) {
+            throw new ElasticsearchParseException(
```
Should we throw `SettingsException` here?

gmarouli replied:
Hm, yes, I can see that being more appropriate :).

andreidan reviewed (commented):
Thanks for working on this Mary

This looks great, I left some minor suggestions

gmarouli commented on docs/reference/indices/get-data-stream.asciidoc:
```diff
@@ -216,6 +220,25 @@ If `true`, the data stream this data stream allows custom routing on write reque
 If `true`, the data stream is created and managed by {ccr} and the local
 cluster can not write into this data stream or change its mappings.
+
+`lifecycle`::
```
It is: `Contains the configuration for the data lifecycle management of this data stream, in experimental:[].` Isn't it clear? I could also put it earlier, if you think it's not good enough.

gmarouli replied:
Sure, then I would like to do the same for now for the `include_defaults` and we can use the the `include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=include-defaults]` and it's not experimental anymore. I can add a todo.

gmarouli replied:
All documentation has been updated.

gmarouli commented on rest-api-spec/src/main/resources/rest-api-spec/api/cluster.get_component_template.json:
```diff
@@ -39,6 +39,10 @@
+      },
+      "include_defaults":{
+        "type":"boolean",
+        "description":"Return relevant default configuration (default: false)"
```
I will incorporate it in a different commit, so I can change the description for the other cases, namely `indices.get_data_stream.json` and `indices.get_index_templates.json`

gmarouli commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverConditions.java:
```diff
@@ -256,6 +257,60 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            } else if (MIN_PRIMARY_SHARD_DOCS_FIELD.getPreferredName().equals(condition)) {
+                builder.addMinPrimaryShardDocsCondition(parseLong(value, condition));
+            } else {
+                throw new ElasticsearchParseException("Unknown condition: '{}'", condition);
```
I will change this one too to `SettingsException`.

gmarouli commented on server/src/main/java/org/elasticsearch/action/admin/indices/template/get/TransportGetComponentTemplateAction.java:
```diff
@@ -50,6 +54,7 @@ public TransportGetComponentTemplateAction(
             GetComponentTemplateAction.Response::new,
             ThreadPool.Names.SAME
         );
+        clusterSettings = DataLifecycle.isEnabled() ? clusterService.getClusterSettings() : null;
```
I was doubting this too :). I will change it.

gmarouli committed:
Review improvements

gmarouli committed:
merge with main

gmarouli commented:
Thanks for the prompt review @andreidan . Along with implementing the feedback from your review I added a test for setting parser in `DataLifecycle` which I removed accidentally during the development of this feature.

elasticsearchmachine commented:
Hi @gmarouli, I've created a changelog YAML for you.

gmarouli committed:
Update docs/changelog/94240.yaml

gmarouli commented:
> Thanks for the prompt review @andreidan . Along with implementing the feedback from your review I added a test for setting parser in DataLifecycle which I removed accidentally during the development of this feature.

Ignore the comment above ^^.... The test was at a different class..... _face_palm_ fixing it

gmarouli committed:
Fix test

gmarouli committed:
Inline the experimental macro.

andreidan commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverConditions.java:
```diff
@@ -256,6 +258,60 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+     * @throws ElasticsearchParseException if the input is invalid, if there are unknown conditions or invalid format values.
+     */
+    public static RolloverConditions parseSetting(String input, String setting) {
+        String[] sConditions = input.split(",");
```
Should we validate against empty (`""`) `input` here?

gmarouli replied:
I will add a specific check, right now it's caught by the next split call.

gmarouli replied:
Added

andreidan commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataLifecycle.java:
```diff
@@ -25,19 +27,52 @@
+     * We require the default rollover conditions to have min_docs set to a non-negative number to avoid empty indices
+     * and to have at least one MAX condition set to ensure that the rollover will be triggered.
+     */
+    static class RolloverConditionsValidator implements Setting.Validator<RolloverConditions> {
```
I'm in two minds about this validator. 
The setting is meant to allow us to make changes, it'll not be something the users will configure. 

I'm tempted to say we shouldn't add this validation, but only validate that the setting is not empty (which should be covered by the `RolloverConditions.parseSetting` ? ) 
What do you think? 


Note: If we do keep it I think it shouldn't be restricted to `min_docs`. The are other `min_*` conditions that could be valid to avoid rolling over empty indices

gmarouli replied:
> The setting is meant to allow us to make changes, it'll not be something the users will configure.
That is true, but I think even we can make a mistake, so avoiding creating empty indices is a safe bet. I see this validation helping the us not blocking us.

> If we do keep it I think it shouldn't be restricted to min_docs. The are other min_* conditions that could be valid to avoid rolling over empty indices
That is true, I did consider it but I had some concerns about the following:
- `min_age` doesn't protect us from empty indices, you can have an index that has been there for 7 days and it can still be empty.
- `min_size` & `min_primary_shard_size` concerns me because an index doesn't have an empty footprint so if you set it up wrongly you might still have an empty index (if I am not mistaken).
- `min_primary_shard_docs` I think is a good addition :).

I would suggest that we keep the validation but we add the `min_primary_shard_docs` as an option too. What do you think? I am considering that there are no use cases that would not want at least one of the two.

gmarouli replied:
Discussed offline. We will remove the checks for `min_*` conditions.

andreidan replied:
I think the whole validator could be removed and leave the RolloverConditions#parser to make sure there's at least one condition, regardless what that is.

andreidan reviewed (approved):
Thanks for iterating on this Mary

Left a few minor suggestions, but this is almost ready to go 🚀

gmarouli committed:
Fix javadoc

Co-authored-by: Andrei Dan <andrei.dan@elastic.co>

gmarouli commented on server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverConditions.java:
```diff
@@ -256,6 +258,60 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
+    /**
+     * Parses a cluster setting configuration, it expects it to have the following format: "condition1=value1,condition2=value2"
+     * @throws ElasticsearchParseException if the input is invalid, if there are unknown conditions or invalid format values.
```
Nice catch!

gmarouli committed:
Merge branch 'main' into dlm-rollover-cluster-setting

gmarouli committed:
Remove validation and capture an empty setting

gmarouli committed:
Merge branch 'main' into dlm-rollover-cluster-setting

gmarouli commented:
@elasticmachine run elasticsearch-ci/part-1
